qa/tasks/util/__init__.py

```diff
@@ -38,6 +39,163 @@ def copy_directory_recursively(from_path, to_remote, to_path=None):
+        assert isinstance(node_roles_list, list), \
+            "node_roles_list is a list"
+        assert node_roles_list, "node_roles_list is not empty"
+        remote = get_remote_for_role(ctx, node_roles_list[0])
+        role_types.append([])
+        if not quiet:
+            logger.debug("Considering remote name {}, hostname {}"
+                         .format(remote.name, remote.hostname))
+        nodes += [remote.hostname]
+        remotes[remote.hostname] = remote
+        remote_lookup_table[remote.hostname] = node_roles_list
+        # inner loop: roles (something like "osd.1" or "c2.mon.a")
+        for role in node_roles_list:
+            # FIXME: support multiple clusters as used in, e.g.,
+            # rgw/multisite suite
+            role_arr = role.split('.')
+            if len(role_arr) != 2:
+                raise ConfigError("Unsupported role ->{}<-"
+                                  .format(role))
+            (role_type, _) = role_arr
+            if role_type not in role_lookup_table:
+                role_lookup_table[role_type] = {}
+            role_lookup_table[role_type][role] = remote.hostname
+            if role_type in cluster_roles:
+                nodes_cluster += [remote.hostname]
+            if role_type in gateway_roles:
+                nodes_gateway += [remote.hostname]
+            if role_type in non_storage_cluster_roles:
+                non_storage_cluster_nodes += [remote.hostname]
+            if role_type == 'osd':
+                nodes_storage += [remote.hostname]
+            if role_type not in role_types[idx]:
+                role_types[idx] += [role_type]
+        idx += 1
+    nodes_cluster = list(set(nodes_cluster))
+    nodes_gateway = list(set(nodes_gateway))
+    nodes_storage = list(set(nodes_storage))
+    nodes_storage_only = []
+    for node in nodes_storage:
+        if node not in non_storage_cluster_nodes:
+            if node not in nodes_gateway:
+                nodes_storage_only += [node]
+    nodes_client_only = list(
+        set(nodes).difference(set(nodes_cluster).union(set(nodes_gateway)))
+        )
+    if not quiet:
+        logger.debug("nodes_client_only is ->{}<-".format(nodes_client_only))
+    assign_vars = [
+        'nodes',
+        'nodes_client_only',
+        'nodes_cluster',
+        'nodes_gateway',
+        'nodes_storage',
+        'nodes_storage_only',
+        'remote_lookup_table',
+        'remotes',
+        'role_lookup_table',
+        'role_types',
+        ]
+    for var in assign_vars:
+        exec("ctx['{var}'] = {var}".format(var=var))
+    ctx['dev_env'] = True if len(nodes_cluster) < 4 else False
+    if not quiet:
+        # report phase
+        logger.info("ROLE INTROSPECTION REPORT")
+        report_vars = assign_vars + ['dev_env']
+        for var in report_vars:
+            logger.info("{} == {}".format(var, ctx[var]))
+
+
+def remote_run_script_as_root(remote, path, data, args=None):
```

Comment:
This can be replaced with remote.sh_file(data, sudo=True)

Reply:
If it uses the `sh` method instead of `run`, then I can't use it.

Reply:
Yeah, `sh_file` uses `sh`, and `sh` has the buffering issue I already explained before.

Reply:
The buffering issue is easy to explain and understand. If I use `run`, each line of output appears in the teuthology.log immediately. If I use `sh`, the output is buffered and is written to the teuthology.log all at once when the remote command completes.

It's really annoying when watching the log in real time when a longer-running command gets buffered like this. Also, I have had cases, in real life, where I was unable to debug a failure because cluster somehow died while running the remote command, and the output never made it into the log.